### PR TITLE
fix(kernel): correctly de-serialize mappings as JSON

### PR DESCRIPTION
**Pull request by RomainMuller**
When mapping data (e.g. a Python `dict`) was passed through a JSON value,
it would not be deserialized correctly and the `$jsii.map` markers would
remain in the JS-visible map.

Fixes the cause of aws/aws-cdk#4957

---

By submitting this pull request, I confirm that my contribution is made under the terms of the [Apache 2.0 license].

[Apache 2.0 license]: https://www.apache.org/licenses/LICENSE-2.0

### CODE REVIEW

File: packages/jsii-kernel/lib/serialization.ts [hunk 1]
```diff
@@ -168,10 +168,33 @@ export const SERIALIZERS: {[k: string]: Serializer} = {
       // Just whatever. Dates will automatically serialize themselves to strings.
       return value;
     },
-    deserialize(value, optionalValue) {
+    deserialize(value, optionalValue, host) {
       // /!\ Top-level "null" will turn to undefined, but any null nested in the value is valid JSON, so it'll stay!
       if (nullAndOk(value, optionalValue)) { return undefined; }
-      return value;
+      if (isWireMap(value)) {
```
Comment (eladb) [inline packages/jsii-kernel/lib/serialization.ts]: can you add a comment explaining why would a value be a "wire map" and what it is?

File: packages/jsii-python-runtime/tests/test_compliance.py [hunk 2]
```diff
@@ -193,8 +193,8 @@ def test_primitiveTypes():
     assert types.date_property == datetime.fromtimestamp(123 / 1000.0, tz=timezone.utc)
 
     # json
-    types.json_property = {"Foo": 123}
-    assert types.json_property.get("Foo") == 123
+    types.json_property = { "Foo": { "bar": 123  } }
```
Comment (eladb) [inline packages/jsii-python-runtime/tests/test_compliance.py]: Would it make sense to do that for java and .net as well?

Comment (eladb) [reply packages/jsii-python-runtime/tests/test_compliance.py]: Did you verify that this indeed fails before the fix is applied?

(Sorry being a bit risk averse)

Comment (RomainMuller) [reply packages/jsii-python-runtime/tests/test_compliance.py]: Oh Java is going to be such fun to write!

Comment (RomainMuller) [reply packages/jsii-python-runtime/tests/test_compliance.py]: Nevermind... Java actually _prohibits_ this particular use-case through static typing!

**Comment by RomainMuller** [comment]
@eladb done!!

**Pull request merged**